flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java

```diff
@@ -528,6 +553,40 @@ public void testOnStartContainerError() throws Exception {
 		}};
 	}
 
+	@Test
+	public void testHandleContainersFromPreviousAttempt() throws Exception {
+		new Context() {{
+			runTest(() -> {
+				Container runningCountainer = mockContainer("container", 1234, 1, resourceManager.getContainerResource());
+				Container newContainer = mockContainer("container", 1234, 2, resourceManager.getContainerResource());
+
+				resourceManager.getTestingNMClientAsync().setGetContainerStatusAsyncBiFunction((containerId, nodeId) -> {
+						if (containerId.equals(runningCountainer.getId())) {
+							return ContainerStatus.newInstance(containerId, ContainerState.RUNNING, "", 0);
+						} else {
+							return ContainerStatus.newInstance(containerId, ContainerState.NEW, "", 0);
+						}
+					});
+
+				List<Container> containersFromPreviousAttempt = new ArrayList<>();
+				containersFromPreviousAttempt.add(newContainer);
+				containersFromPreviousAttempt.add(runningCountainer);
+				RegisterApplicationMasterResponse testResponse = RegisterApplicationMasterResponse.newInstance(
+					resourceManager.getContainerResource(),
+					resourceManager.getContainerResource(),
+					Collections.emptyMap(),
+					null,
+					containersFromPreviousAttempt,
+					"",
+					Collections.emptyList());
+				resourceManager.getContainersFromPreviousAttempts(testResponse);
+
+				verify(mockResourceManagerClient).releaseAssignedContainer(newContainer.getId());
+				assertEquals(1, resourceManager.getWorkerNodeMap().size());
```

Comment:
In addition to the size, we can also verify the worker in the worker not map is the running container.